include/pagmo/algorithms/cstrs_self_adaptive.hpp

```diff
@@ -0,0 +1,783 @@
+        assert(pop.size() >= 4u);                   // Population cannot contain less than 3 individuals
+
+        // We assign the naked pointer The pointer will be immutable (as in its never changed afterwards)
+        m_pop_ptr = &pop;
+        // Update all data members and init the cache
+        update();
+    }
+
+    // The bounds are unchanged
+    std::pair<vector_double, vector_double> get_bounds() const
+    {
+        return m_pop_ptr->get_problem().get_bounds();
+    }
+
+    /// The fitness computation
+    vector_double fitness(const vector_double &x) const
+    {
+        double solution_infeasibility;
+        vector_double f(1, 0.);
+
+        // 1 - We check if the decision vector is already in the reference population and return that or recompute.
+        auto it_f = m_fitness_map.find(x);
+        if (it_f != m_fitness_map.end()) {
+            f[0] = it_f->second[0];
+            solution_infeasibility = compute_infeasibility(it_f->second);
+        } else { // we have to compute the fitness (this will increase the feval counter in the ref pop problem )
+            auto fit = m_pop_ptr->get_problem().fitness(x);
+            f[0] = fit[0];
+            solution_infeasibility = compute_infeasibility(fit);
+            m_fitness_map[x] = fit;
+        }
+        // 2 - Then we apply the penalty
+        if (solution_infeasibility > 0.) {
+            double inf_tilde = solution_infeasibility;
+            if (m_i_hat_up != m_i_hat_down) {
+                inf_tilde = (solution_infeasibility - m_i_hat_down) / (m_i_hat_up - m_i_hat_down);
+            } else {
+                inf_tilde = solution_infeasibility; // This will trigger, for example, when the whole population is
+                                                    // feasible.
+            }
+            // apply penalty 1 only if necessary. This penalizes infeasible solutions so that their objective
+            // values cannot be much better than that of the best solution.
+            if (m_apply_penalty_1) {
+                f[0] += inf_tilde * (m_f_hat_down[0] - m_f_hat_up[0]);
+            }
+            // apply penalty 2 NOTE: uses (2. * inf_tilde) correcting what seems a clear bug in pagmo legacy
+            // This penalizes all infeasible solutions exponentially with their infeasibility
+            f[0] += m_scaling_factor * std::abs(f[0]) * ((std::exp(2. * inf_tilde) - 1.) / (std::exp(2.) - 1.));
+        }
+        return f;
+    }
+
+    // Call to this method updates all the members that are used to penalize the objective function
+    // As the penalization algorithm depends heavily on the ref population this method takes care of
+    // updating the necessary information. It also builds the hash map used to avoid unecessary fitness
+    // evaluations. We exclude this method from the test as all of its corner cases are difficult to trigger
+    // and test for correctness
+    // LCOV_EXCL_START
```

Comment:
I feel like this is cheating. Normally we do this only for 3rd party code, which IMO is fine. But for our code, if it's really so difficult to cover all the test, at least we should pay the "iron price". Let's not hide the dust under the carpet :)